### PR TITLE
fix: add workflow outputs to scope for arguments

### LINKED ISSUE #11610: Cannot access output global artifacts in exit handler after upgrading to v3.4.6+
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

Whenever we are calling the global output artifact in the exithandler Its failing with the error:

Bad Request: templates.exit-handler.steps failed to resolve ```{{workflow.outputs.artifacts.output-result-car}}```

This error was introduced after upgrading to the v3.4.6,. Also have tested with v3.4.7, v3.4.8, v3.4.9, v3.4.10 same issue persist. 

After downgrading to v3.4.5  same workflow is working without any issue

### Version

v3.4.6

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: exit-handler-with-artifacts
spec:
  onExit: exit-handler
  entrypoint: main
  arguments:
    parameters:
    - name: params
      value: |
        [
          { "variable": "car"},
          { "variable": "bike"}
        ]
  templates:
    - name: main
      steps:
        - - name: step-1
            template: output
			arguments:
			  parameters:
			  - name: variable
			    value: "{{item.variable}}"
            withParam: "{{workflow.parameters.params}}" 
    
    - name: output
	  inputs:
        parameters:
        - name: variable
      script:
        image: python:alpine3.6
        command: [sh, -c]
        args: ["echo $variable > /result.txt "]
      outputs:
        artifacts:
          - name: result-{{inputs.parameters.variable}}
            path: /result.txt
            globalName: output-result-{{inputs.parameters.variable}}

    - name: printer
      inputs:
        artifacts:
          - name: message-bike
            path: /tmp/message-bike
          - name: message-car
            path: /tmp/message-car
      container:
        image: python:alpine3.6
        command: [sh, -c]
        args: ["cat /tmp/message.*"]
			
    - name: exit-handler
	  steps:
	    - - name: printer
			template: printer
			arguments:
			  artifacts:
			  - name: message-car
				from: "{{workflow.outputs.artifacts.output-result-car}}"          
			  - name: message-bike
				from: "{{workflow.outputs.artifacts.output-result-bike}}" 
```
```


### Logs from the workflow controller

```text
NA
```

### Logs from in your workflow's wait container

```text
NA
```

### PR DESCRIPTION
**Pull request by rohankmr414**
Fixes #11610

### Modifications

Added workflow global outputs to scope for arguments

### Verification

Tested locally

### CODE REVIEW

**Review by isubasinghe — APPROVED** [review]
LGTM, is it possible to add a test for this ? Perhaps in e2e

File: workflow/validate/validate.go [hunk 1]
```diff
@@ -1374,7 +1379,7 @@ func (ctx *templateValidationCtx) validateDAG(scope map[string]interface{}, tmpl
 		ancestry := common.GetTaskAncestry(dagValidationCtx, task.Name)
 		for _, ancestor := range ancestry {
 			ancestorTask := dagValidationCtx.GetTask(ancestor)
-			resolvedTmpl := resolvedTemplates[ancestor]
+			resolvedTmpl = resolvedTemplates[ancestor]
```
Comment (Joibel) [inline workflow/validate/validate.go:1382]: This change doesn't actually do anything but is probably a wrong change and at minimum would be best reverted.

`resolvedTmpl` is not used outside of this for loop in the outer scope **after** this loop. So the fact that this variable used to shadow the outer `resolvedTmpl` but doesn't any more doesn't actually change the behaviour. However - this value of resolvedTmpl is definitely _not_ useful outside of this loop, and the formerly shadowed version might be. The right fix is to use a different name for this variable inside this loop.

Comment (rohankmr414) [reply workflow/validate/validate.go:1382]: reverting this

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -3054,6 +3054,11 @@ func (tmpl *Template) HasOutput() bool {
 	return tmpl.Container != nil || tmpl.ContainerSet.HasContainerNamed("main") || tmpl.Script != nil || tmpl.Data != nil || tmpl.HTTP != nil || tmpl.Plugin != nil
 }
 
+// HasArguments returns true if the template can have arguments.
+func (tmpl *Template) HasArguments() bool {
```
Comment (Joibel) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:3058]: I'd prefer if this was `CanHaveArguments` as it's slightly misleadingly named as is.

**Comment by Joibel** [comment]
@rohankmr414 shall I'll take over this PR

**Comment by tooptoop4** [comment]
☂️

**Pull request closed without merging**